### ISSUE TITLE
Pagination with bookmark

### ISSUE THREAD

**Issue by will-moore** [ome/omero_search_engine#47]
Thanks for the example at https://github.com/ome/omero_search_engine/blob/main/examples/search_with_bookmark_paging_using_submitquery.py

If I understand correctly, subsequent calls to `/api/v1/resources/submitquery` with a bookmark will return a series of result pages (without needing to include the page number each time).

However, it would be nice and maybe a bit more robust if you could request a specific page of results..
E.g. If I wanted to show separate pages of search results and allow a user to go forwards and backwards between them.
Or, even if a page is taking a while to respond, and the user cancels and refreshes - then they won't get the same page again but will get the next page.

I don't know how this is implemented on the back end, so there may be a good reason for this behaviour. Also, I'm probably going to display this like an "infinite scroll" which won't allow a user to go back and forth (or jump to the last page), so it will likely work OK for me, but I can imagine other API users requesting indexed pagination.

Also, I was wondering if it's really necessary to include the original query each time you load a page with the bookmark?
It's only a bit more code to do this each time, but is it actually used by the back end? What if I submit a different query each time?

Thanks

**Comment by will-moore** [comment]
Also I'm wondering why `bookmark` is a list with a single item in it?
```"bookmark": [107448]```

**Comment by will-moore** [comment]
Just started to use the `bookmark` now, and I noticed that when I ask for the 2nd page of results, using e.g. `bookmark: [3681511]`, the bookmark that is returned from that query is different, e.g. `[3745832]`.
Since I was updating my local value of `bookmark` with the returned value each time, this led to errors.
When I ignored the subsequent values and used the first value for all requests, it works OK. Not sure if this is a bug or intended feature?

**Comment by khaledk2** [comment]
The query should be sent along with a bookmark when calling the next page.
The results contains also the total number of pages and the total number of images.
The user should use the bookmark which is returned with the results to access the next page.
The bookmark is an array with one item, it is the format that the Easticsearch accepts, maybe it is a good idea to use a single int instead and the searchengine will deal with the Easticsearch query format.
The client can use the bookmark to explore the previous page or the next page.
The client can not call a specific page, maybe it can be done after getting all the pages and mapping the bookmark with the relevant page.
Sorry, the page should not return with the query results, it is something the client should manage, I will remove that.

**Comment by will-moore** [comment]
The `page` increments from 1 to 2, and is useful to know which page you are on (so you know if there are more results available). If not `page` then it would be good to have another flag to know when there are no more pages available.

If I were showing pages like `/results/?bookmark=12345` I could show a `next` button on that page if I know there are more results. I could also show a `previous` button, but only if the server returned me a "previous" bookmark (instead of just the "next" bookmark). If the server is already keeping track of pagination, then it's easier to share that state with the client, instead of needing the client to keep track of it too.

**Comment by khaledk2** [comment]
The search is stateless, so it does not keep track of pages.
The server knows about the next and current bookmark but it is not aware of the previous bookmark.
Although the server does not send all the results in one call, if it is more than 1000, it sends the total number of items inside the results (inside all the pages) and the total number of pages.
We may implement something to keep track of pages.
We may add a part to the results; it should be sent back with the query when calling the next page. The server will update it and send it again to the client with the results.
It should be something like that:

```
   "paggening":{
      "current page":4,
      "page_records":[
         {
            "page":2,
            "bookmark":234556
         },
         {
            "page":3,
            "bookmark":455678
         },
         {
            "page":4,
            "bookmark":678954
         }
      ]
   }
```

**Comment by will-moore** [comment]
That would be great, as I will then know when the last page has been loaded. 👍 .
`"paggening"` should probably be "pagination"?

**Comment by khaledk2** [comment]
This has been implemented on PR #65
